File: ts/vbcode/VBparser.ts
namespace vscode {

    /** 
     * The VB code syntax token generator
    */
    export class VBParser {

        private code: tokenStyler = new tokenStyler();
        private escapes = {
            string: false,
            comment: false,
            keyword: false // VB之中使用[]进行关键词的转义操作
        };
        private token: string[] = [];

        /** 
         * @param chars A chars enumerator
        */
        public constructor(private chars: Pointer<string>) {
        }

        /**
         * Get source file document highlight result
        */
        public GetTokens(): tokenStyler {
            while (!this.chars.EndRead) {
                this.walkChar(this.chars.Next);
            }

            return this.code;
        }

        private static peekNextToken(chars: Pointer<string>, allowNewLine: boolean = false): string {
            var i: number = 0;
            var c: string = null;

            while ((c = chars.Peek(i++)) == " " || c == "\n") {
                if ((c == "\n") && !allowNewLine) {
                    break;
                }
            }

            return c;
        }

        private get isKeyWord(): boolean {
            return VBKeywords.indexOf(this.token.join("")) > -1;
        }
        private get isAttribute(): boolean {
            var token = this.token;
            var haveTagEnd = token[token.length - 1] == ">" || this.chars.Peek(-1) == "("; // token[token.length - 1] == "(";

            return token[0] == "&lt;" && haveTagEnd;
        }
        private endToken() {
            var code = this.code, token = this.token;

            if (this.isAttribute) {
                // 自定义属性需要一些额外的处理
                // 不渲染符号，只渲染单词
                code.append(token[0]);

                if (token[token.length - 1] == ">") {
                    code.attribute($ts(token).Skip(1).Take(token.length - 2).JoinBy(""));
                    code.append(token[token.length - 1]);
                } else {
                    code.attribute($ts(token).Skip(1).Take(token.length - 1).JoinBy(""));
                }

            } else if (code.LastNewLine && token[0] == "#") {
                code.directive(token.join(""));
            } else {
                // 结束当前的单词
                var word: string = token.join("");

                if (code.LastDirective) {
                    code.directive(word);
                } else if (VBKeywords.indexOf(word) > -1) {
                    // 当前的单词是一个关键词
                    code.keyword(word);
                } else if (code.LastTypeKeyword) {
                    if (code.LastAddedToken == "Imports") {
                        // Imports xxx = yyyy
                        if (VBParser.peekNextToken(this.chars) == "=") {
                            code.type(word);
                        } else {
                            code.append(word);
                        }
                    } else if (word == "(") {
                        code.append(word);
                    } else {
                        code.type(word);
                    }
                } else {
                    code.append(word);
                }
            }

            this.token = [];
        }

        /** 
         * 处理当前的这个换行符
        */
        private walkNewLine() {
            // 是一个换行符
            if (this.escapes.comment) {
                // vb之中注释只有单行注释，换行之后就结束了                    
                this.code.comment(this.token.join("").replace("<", "&lt;"));
                this.escapes.comment = false;
                this.token = [];
            } else if (this.escapes.string) {
                // vb之中支持多行文本字符串，所以继续添加
                // this.token.push("<br />");
                this.code.string(this.token.join(""));
                this.code.appendLine();
                this.token = [];
            } else {
                // 结束当前的token
                this.endToken();
                this.code.appendLine();
            }
        }

        private walkStringQuot() {
            // 可能是字符串的起始
            if (!this.escapes.string) {
                this.escapes.string = true;
                this.endToken();
                this.token.push('"');
            } else if (this.escapes.string) {
                // 是字符串的结束符号
                this.escapes.string = false;
                this.token.push('"');
                this.code.string(this.token.join(""));
                this.token = [];
            }
        }

        private walkChar(c: string): void {
            var escapes = this.escapes;
            var code = this.code;

            if (c == "\n") {
                this.walkNewLine();

            } else if (this.escapes.comment) {
                // 当前的内容是注释的一部分，则直接添加内容
                this.token.push(c);

                // 下面的所有代码都是处理的非注释部分的内容了
                // 代码注释部分的内容已经在处理换行符和上面的代码之中完成了处理操作

            } else if (c == '"') {
                this.walkStringQuot();

            } else if (c == "'") {
                if (!escapes.string) {
                    // 是注释的起始
                    escapes.comment = true;
                    this.endToken();
                    this.token.push(c);
                } else {
                    this.token.push(c);
                }
            } else if (c == " " || c == "\t") {
                // 使用空格进行分词
                if (!escapes.string) {
                    this.endToken();
                    code.append(c);

                    // 是字符串的一部分
                } else if (c == " ") {
                    this.token.push("&nbsp;");
                } else {
                    // 是一个TAB
                    // 则插入4个空格
                    for (var i: number = 0; i < 4; i++) {
                        this.token.push("&nbsp;");
                    }
                }
            } else if (c in delimiterSymbols) {
                // 也会使用小数点进行分词
                if (!escapes.string) {
                    if (c == "(") {
                        //if (this.isKeyWord) {
                        this.endToken();
                        this.token.push("(");
                        this.endToken();
                        //} else {
                        //    this.endToken();
                        //    this.token.push("(");
                        //    this.endToken();
                        //}
                    } else {
                        this.endToken();
                        code.append(c);
                    }
                } else {
                    this.token.push(c);
                }
            } else if (c == "<" || c == "&") {
                this.token.push(Strings.escapeXml(c));
            } else {
                this.token.push(c);
            }
        }
    }
}

File: vbcode/vbcode/VBParser.vb
Imports Microsoft.VisualBasic.Emit.Marshal
Imports Microsoft.VisualBasic.Language
Imports Microsoft.VisualBasic.Text

Public Class VBParser

    Dim code As New tokenStyler
    Dim escapes As New escapes
    Dim token As New List(Of Char)

    Dim chars As Pointer(Of Char)

    Public ReadOnly Property isKeyword As Boolean
        Get
            Return token.JoinBy("") Like Highlighter.VBKeywords
        End Get
    End Property

    Public ReadOnly Property isAttribute As Boolean
        Get
            If token > 0 Then
                Dim haveTagEnd = token.Last = ">"c OrElse chars.Value(-1) = "("c
                Return token(Scan0) = "<" AndAlso haveTagEnd
            Else
                Return False
            End If
        End Get
    End Property

    Sub New(chars As Pointer(Of Char))
        Me.chars = chars
    End Sub

    Public Function getTokens() As tokenStyler
        Do While Not chars.EndRead
            Call walkChar(++chars)
        Loop

        If token > 0 Then
            Call walknewline
        End If

        Call code.flush()
        Return code
    End Function

    Private Shared Function peekNextToken(chars As Pointer(Of Char), Optional allowNewline As Boolean = False) As String
        Dim i As VBInteger = 1
        Dim c As Value(Of Char) = ""

        Do While ((c = chars.RawBuffer(++i)) = " "c) OrElse c Like ASCII.CR OrElse c Like ASCII.LF
            If c Like ASCII.CR OrElse c Like ASCII.LF AndAlso Not allowNewline Then
                Exit Do
            End If
        Loop

        Return c
    End Function

    Private Sub endToken()
        If token = 0 Then
            Return
        End If

        If isAttribute Then
            code.append(token(0))

            If token.Last = ">"c Then
                code.attribute(token.Skip(1).Take(token.Count - 2).JoinBy(""))
                code.append(token.Last)
            Else
                code.attribute(token.Skip(1).Take(token.Count - 1).JoinBy(""))
            End If
        ElseIf code.LastNewLine AndAlso token(Scan0) = "#"c Then
            code.directive(token.CharString)
        Else
            Dim word As New String(token)

            If code.LastDirective Then
                code.directive(word)
            ElseIf word Like VBKeywords Then
                code.keyword(word)
            ElseIf code.LastTypeKeyword Then
                If code.LastAddedToken = "Imports" Then

                    If VBParser.peekNextToken(chars) = "=" Then
                        code.type(word)
                    Else
                        code.append(word)
                    End If
                ElseIf word = "(" Then
                    code.append(word)
                Else
                    code.type(word)
                End If
            Else
                code.append(word)
            End If
        End If

        token *= 0
    End Sub

    Private Sub walkNewline()
        If escapes.comment Then
            code.comment(token.CharString)
            escapes.comment = False
            token *= 0
        ElseIf escapes.string Then
            code.string(token.CharString)
            code.appendLine()
            token *= 0
        Else
            endToken()
            code.appendLine()
        End If
    End Sub

    Private Sub walkStringQuot()
        If Not escapes.string Then
            escapes.string = True
            endToken()
            token += """"c
        ElseIf escapes.string Then
            escapes.string = False
            token += """"c
            code.string(token.CharString)
            token *= 0
        End If
    End Sub

    Private Sub addToken(c As Char)
        If c = " "c Then
            token += " "
        ElseIf c = ASCII.TAB Then
            token += {" "c, " "c, " "c, " "c}
        Else
            token += c
        End If
    End Sub

    Private Sub walkChar(c As Char)
        If c = ASCII.LF Then
            walkNewline()
        ElseIf c = ASCII.CR Then
            ' do nothing
        ElseIf escapes.comment Then
            addToken(c)
        ElseIf c = """"c Then
            walkStringQuot()
        ElseIf c = "'"c Then
            If Not escapes.string Then
                escapes.comment = True
                endToken()
                token += c
            Else
                token += c
            End If
        ElseIf c = " "c OrElse c = ASCII.TAB Then
            If Not escapes.string Then
                endToken()
                code.append(c)
            Else
                addToken(c)
            End If
        ElseIf c Like delimiterSymbols Then
            If Not escapes.string Then
                If c = "("c Then
                    endToken()
                    token += "("c
                    endToken()
                Else
                    endToken()
                    code.append(c)
                End If
            Else
                token += c
            End If
        ElseIf c = "<"c OrElse c = "&"c Then
            token += c
        Else
            token += c
        End If
    End Sub

End Class

Public Class escapes
    Public [string] As Boolean = False
    Public comment As Boolean = False
    ''' <summary>
    ''' VB之中使用[]进行关键词的转义操作
    ''' </summary>
    Public keyword As Boolean = False
End Class